src/common/node/promise.ts

```diff
@@ -22,6 +22,17 @@ export class PromiseUtil {
         return this.retryAsyncIteration(operation, condition, maxRetries, 0, delay, failure);
     }
 
+    public reduce<T>(sources: T[]|Q.Promise<T[]>, generateAsyncOperation: (value: T) => Q.Promise<void>): Q.Promise<void> {
```

Comment:
Why not also take a generic parameter `U` for the return type instead of enforcing null?